### PR TITLE
Added syntax support for JSON files

### PR DESCRIPTION
**Pull request by c6c534e825f4a3d41ede3e67473187d1**
Looks cleaner than the default syntax `language_js` gives

Example:
2023-11-05-234949_380x215_scrot

### CODE REVIEW

File: plugins/language_json.lua
Comment (jgmdev) [inline plugins/language_json.lua:1]:
```suggestion
-- mod-version:3 priority:90
```

Since language_js.lua already supports .json files the priority of this plugin should be higher than default so it gets loaded and applied before language_js.

Comment (c6c534e825f4a3d41ede3e67473187d1) [reply plugins/language_json.lua]: Didn't knew that was possible, cheers!

Maybe the `.json` file extension should be removed from `language_js` aswell, or should it be kept as backup?

Comment (jgmdev) [reply plugins/language_json.lua]: I guess it should be kept to provide some json support on default installs but the maintainers will tell.

File: plugins/language_json.lua
Comment (Guldoman) [inline plugins/language_json.lua:22]: I wonder if those should be removed, as they're invalid JSON.

File: plugins/language_json.lua
Comment (Guldoman) [inline plugins/language_json.lua:12]: These should probably allow for spaces between the key and `:`.

Also I think you can avoid escaping `"` here.

Are numeric keys allowed in JSON?

File: plugins/language_json.lua
Comment (Guldoman) [inline plugins/language_json.lua:13]: I wonder if this should be adapted to disallow multiline strings.

File: plugins/language_json.lua
Comment (c6c534e825f4a3d41ede3e67473187d1) [reply plugins/language_json.lua:12]: You're right, numeric keys are not allowed according to the JSON spec. Got them confused with how python handles them.

File: plugins/language_json.lua
Comment (c6c534e825f4a3d41ede3e67473187d1) [reply plugins/language_json.lua:22]: I only added them because I've seen it in real world examples, but yeah as they're invalid JSON I think it's best to strip them, or let the user re-add them incase it's neccesary

**Comment by c6c534e825f4a3d41ede3e67473187d1** [comment]
Just noticed another issue, keys that contain `-` are highlighted incorrectly:
2023-11-06-135957_348x487_scrot

File: plugins/language_json.lua
Comment (c6c534e825f4a3d41ede3e67473187d1) [reply plugins/language_json.lua]: Coming  back to this after doing some testing, it seems like `language_js` is loaded before `language_json`, but after removing `priority:90` it works fine again

File: plugins/language_json.lua
Comment (c6c534e825f4a3d41ede3e67473187d1) [reply plugins/language_json.lua:12]: I've got it to ignore strings like:
```
"foo": "

bar",
```

Should it also ignore `"""`, `''` and the like?

File: plugins/language_json.lua
Comment (c6c534e825f4a3d41ede3e67473187d1) [reply plugins/language_json.lua:11]: Yup, added it to the pattern

**Comment by c6c534e825f4a3d41ede3e67473187d1** [comment]
Excuse the commit mess, I am not 100% awake and in a rush aswell

**Comment by Guldoman** [comment]
> Excuse the commit mess, I am not 100% awake and in a rush aswell

Don't worry about it, it'll all get squashed anyways.

File: plugins/language_json.lua
Comment (Guldoman) [reply plugins/language_json.lua]: I mean between the end of the string and `:`, like `"blabla"   :`.

File: plugins/language_json.lua
Comment (Guldoman) [inline plugins/language_json.lua:1]: `110` would probably work better I think.

File: plugins/language_json.lua
Comment (Guldoman) [inline plugins/language_json.lua:11]: I wonder if we could use the second pattern for the initial part of the first, like:
```suggestion
    { pattern = '"[^\n]*"%s*:', type = "keyword" }, -- key
    { pattern = '"[^\n]*"', type = "string" }, -- value
```
So that it would recognize keys of any kind.

File: plugins/language_json.lua
Comment (c6c534e825f4a3d41ede3e67473187d1) [reply plugins/language_json.lua]: Ahh, `[ ]*` would do the trick as using `\w` results in getting errors

File: plugins/language_json.lua
Comment (c6c534e825f4a3d41ede3e67473187d1) [reply plugins/language_json.lua:1]: Yeah, that works

File: plugins/language_json.lua
Comment (Guldoman) [reply plugins/language_json.lua:11]: In the end the only issue is that it might break when trying to escape strings, but we can live with that for now.

**Review by Guldoman — APPROVED** [review]
Looks good. Will wait a bit, then merge.

**Comment by jgmdev** [comment]
Maybe you could also add [.cjson](https://www.webcomand.com/docs/developer_guide/repositories/storage_engines/file_system/cjson/index.html) support which is same as `.json` but adds multiline comments support `/**/` and  single comment lines `//`

**Comment by c6c534e825f4a3d41ede3e67473187d1** [comment]
Yeah, I suppose that shouldn't be too hard to implement

**Comment by c6c534e825f4a3d41ede3e67473187d1** [comment]
Maybe we should eventually add syntax support for cJSON itself in a new file, as it's a bit more than just comments?

**Comment by Guldoman** [comment]
What else does it add?

**Comment by c6c534e825f4a3d41ede3e67473187d1** [comment]
It has a strict format, and a cJSON payload must include objects such as `type` and `version`.
Here's an example:
```
/**
 * Contacts - Some contacts with photos.
 */
{
    "type": "COMAND",
    "version": "1.0",
    "title": "Contacts",
    "description": "Some contacts with photos.",
    "source": "https://comandsystems.com/contacts",
    "time": "2013-10-16T15:39:25+0400",
    "contents": [
        {
            "ID": 13,
            "OID": 2458,
            "UUID": "550e8400-e29b-41d4-a716-446655440013",
            "Type": {"UUID": "550e8400-e29b-41d4-a716-446655440123"},
            "Name": "Mary Smith",
            "EmergencyContact": {"OID": 2460},
            "PhotoFilename": "profile.jpg",
            "Photo": {"type": "url", "value": "files/1.jpg"},
            "Phones": [
                {
                    "ID": 1,
                    "OID": 2478,
                    "UUID": "550e8400-e29b-41d4-a716-446655442323",
                    "Type": {"UUID": "550e8400-e29b-41d4-a716-446655441122"},
                    "Number": "555-555-5555",
                    "Description": "Mobile"
                }
            ]
        },
        {
            "ID": 15,
            "OID": 2460,
            "UUID": "550e8400-e29b-41d4-a716-446655440015",
            "Type": {"UUID": "550e8400-e29b-41d4-a716-446655440123"},
            "Name": "John Smith",
            "EmergencyContact": {"OID": 2458}, // reference to Mary Smith
            "PhotoFilename": "mug.jpg",
            "Photo": {"type": "url", "value": "files/2.jpg"}
        }
    ]
}
```

**Comment by Guldoman** [comment]
Mmm, yeah would make sense to move that in a different language, and highlight those special keys.

Btw I think we're conflating cjson and jsonc. Where the latter is the one that "simply" adds comments (maybe something more, I'd have to check).

**Comment by jgmdev** [comment]
Found some issue while working on a manifest file:

json-issue

File: plugins/language_json.lua
Comment (Guldoman) [inline plugins/language_json.lua:17]:
```suggestion
    { pattern = '"[^\n]-"%s*:', type = "keyword" }, -- key
    { pattern = '"[^\n]-"', type = "string" }, -- value
```

Comment (Guldoman) [reply plugins/language_json.lua:17]: Also, to avoid highlighting the `:`:
```suggestion
    { pattern = '"[^\n]*"%s*():', type = { "keyword", "normal"} }, -- key
    { pattern = '"[^\n]*"', type = "string" }, -- value
```

**Comment by jgmdev** [comment]
Regular expressions to the rescue:

```lua
    { regex = [["(?:[^"\\]|\\.)*"\s*:]], type = "keyword" }, -- key
    { regex = [["(?:[^"\\]|\\.)*"]], type = "string" }, -- value
```

json-with-escape


Regex recipe from: https://stackoverflow.com/a/249937

**Comment by c6c534e825f4a3d41ede3e67473187d1** [comment]
> Regular expressions to the rescue:
> 
> ```lua
>     { regex = [["(?:[^"\\]|\\.)*"\s*:]], type = "keyword" }, -- key
>     { regex = [["(?:[^"\\]|\\.)*"]], type = "string" }, -- value
> ```

Does this also fix the highlighting of `:`?

**Comment by Guldoman** [comment]
> > Regular expressions to the rescue:
> > ```lua
> >     { regex = [["(?:[^"\\]|\\.)*"\s*:]], type = "keyword" }, -- key
> >     { regex = [["(?:[^"\\]|\\.)*"]], type = "string" }, -- value
> > ```
> 
> Does this also fix the highlighting of `:`?

It does not. This will tho:
```lua
    { regex = [["(?:[^"\\]|\\.)*"()\s*:]], type = { "keyword", "normal" } }, -- key
    { regex = [["(?:[^"\\]|\\.)*"]], type = "string" }, -- value
```

**Comment by c6c534e825f4a3d41ede3e67473187d1** [comment]
> ```lua
> { regex = [["(?:[^"\\]|\\.)*"()\s*:]], type = { "keyword", "normal" } }, -- key
>     { regex = [["(?:[^"\\]|\\.)*"]], type = "string" }, -- value
> ```

Just added this, and it seems to be working good! I'll add it in a sec

File: plugins/language_json.lua
Comment (Guldoman) [inline plugins/language_json.lua:7]: I'd also add `.jsonc`.

**Review by Guldoman — APPROVED** [review]
Looks good, thank you!
Only one last nit, and after that's fixed I'll wait a bit more and merge.

**Comment by Guldoman** [comment]
Thank you!

**Pull request merged**
